packages/core/src/controllers/events.ts

```diff
@@ -0,0 +1,200 @@
+    this.toPersist = true;
+  };
+
+  private setEventListeners = () => {
+    this.core.heartbeat.on(HEARTBEAT_EVENTS.pulse, async () => {
+      if (this.toPersist) await this.persist();
+      // cleanup events older than EVENTS_STORAGE_CLEANUP_INTERVAL
+      this.events.forEach((event) => {
+        if (
+          fromMiliseconds(Date.now()) - fromMiliseconds(event.timestamp) >
+          EVENTS_STORAGE_CLEANUP_INTERVAL
+        ) {
+          this.events.delete(event.eventId);
+          this.toPersist = true;
+        }
+      });
+    });
+  };
+
+  private setMethods = (eventId: string) => {
+    return {
+      addTrace: (trace: string) => this.addTrace(eventId, trace),
+      setError: (errorType: string) => this.setError(eventId, errorType),
+    };
+  };
+
+  private addTrace = async (eventId: string, trace: string) => {
+    const event = this.events.get(eventId);
+    if (!event) return;
+    return await new Promise<void>((resolve) => {
+      event.props.properties.trace.push(trace);
+      this.events.set(eventId, event);
+      this.toPersist = true;
+      resolve();
+    });
+  };
+
+  private setError = async (eventId: string, errorType: string) => {
+    const event = this.events.get(eventId);
+    if (!event) return;
+    return await new Promise<void>((resolve) => {
+      event.props.type = errorType;
+      event.timestamp = Date.now();
+      this.events.set(eventId, event);
+      this.toPersist = true;
+      resolve();
+    });
+  };
+
+  private persist = async () => {
+    await this.core.storage.setItem(this.storageKey, Array.from(this.events.values()));
+    this.toPersist = false;
+  };
+
+  private restore = async () => {
+    try {
+      const events =
+        (await this.core.storage.getItem<EventClientTypes.Event[]>(this.storageKey)) || [];
+      if (!events.length) return;
+      events.forEach((event) => {
+        this.events.set(event.eventId, {
+          ...event,
+          ...this.setMethods(event.eventId),
+        });
+      });
+    } catch (error) {
+      this.logger.warn(error);
+    }
+  };
+
+  private submit = async () => {
+    if (!this.telemetryEnabled) return;
+
+    if (this.events.size === 0) return;
+
+    const eventsToSend = [];
+    // exclude events without type as they can be considered `in progress`
+    for (const [_, event] of this.events) {
+      if (event.props.type) {
+        eventsToSend.push(event);
+      }
+    }
+
+    if (eventsToSend.length === 0) return;
+
+    try {
+      const response = await fetch(EVENTS_CLIENT_API_URL, {
+        method: "POST",
+        body: JSON.stringify(eventsToSend),
+        headers: {
+          "x-project-id": `${this.core.projectId}`,
+          "x-sdk-type": "events_sdk",
+          "x-sdk-version": `js-${RELAYER_SDK_VERSION}`,
+        },
+      });
+      if (response.ok) {
+        for (const event of eventsToSend) {
+          this.events.delete(event.eventId);
+          this.toPersist = true;
```

Comment:
I'm worry of race conditions with this variable being mutated in various places.

Reply:
`eventsToSend` shouldn't mutate because it is declared before the http call https://github.com/WalletConnect/walletconnect-monorepo/blob/fa4d6c9ddd83fe79df2dc1093831ba5bd814ccd5/packages/core/src/controllers/events.ts#L170
and even if `this.events` mutate during that, it wont remove the new values